File: go/string/longest-substring-without-repeating-characters.go
package string

import "strings"

func LengthOfLongestSubstring(s string) int {
	length := len(s)
	if length <= 1 {
		return length
	}

	max := 1

	right := 0

	route := ""

	for right <= length-1 {
		curChar := string(s[right])
		index := strings.Index(route, curChar)
		if index >= 0 {
			route = route[index+1:]
		}
		right++
		route = route + curChar
		max = maxFunc(max, len(route))
	}
	return max
}

func LengthOfLongestSubstring2(s string) int {
	//字符串的总长度
	length := len(s)
	if length <= 1 {
		return length
	}

	//最大子串长度
	max := 1
	//右指针
	right := 1
	//左指针
	left := 0

	for right <= length-1 {
		//当前字符
		curChar := s[right]
		//默认在[left,right)不存在与s[right]相同的字符
		index := -1
		//在[left,right)中找是否有与s[right]相同的字符
		for i := left; i < right; i++ {
			if curChar == s[i] {
				index = i
			}
		}
		//如果找到了,则left指针指向这个字符的下一个,此时由于子串长度缩短,所以不更新最大子串长度
		if index >= 0 {
			left = index + 1
		} else { //如果没找到,则更新最大子串长度
			max = maxFunc(max, right-left+1)
		}
		right++
	}
	return max
}

func maxFunc(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

File: string/longest-substring-without-repeating-characters.php
<?php

class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function lengthOfLongestSubstring($s) {
        if (strlen($s) <= 1) {
            return strlen($s);
        }
        //右指针
        $right = 0;
        //最小长度,用来记录唯一串的最大长度
        $max = 1;
        //临时的唯一串
        $route = '';
        //如果右指针指的元素存在且不存在临时唯一串里面
        while (isset($s[$right])) {
            $pos = strpos($route, $s[$right]);
            //如果唯一串里面已经存在这个元素,则把唯一串中的这个元素及这个元素之前的都剔除,然后把当前字符放进唯一串
            if ($pos !== false) {
                $route = substr($route, $pos + 1);
            }
            //当前字符放进字符串尾部
            $route .= $s[$right];
            //计算当前串的长度
            $max = max($max, strlen($route));
            //指针右移
            $right++;
        }
        return $max;
    }
}

echo (new Solution())->lengthOfLongestSubstring('aaw');